### DISCUSSION TITLE
Typescript: How do I define getInitialProps in a custom _app

### DISCUSSION THREAD

**Discussion by arsen** [vercel/next.js#11900] [Help]
Hi,
Could someone please help me?
I'm trying to define a custom _app component in TypeScript.
Which should read cookies from getInitialProps function and adjust the Layout based on cookies.

My question is what should I return from getInitialProps, 
what are the correct typescript types for MyApp.getInitialProps = async (ctx)

Here is my example code:

```typescript
const MyApp = (props: AppProps) => {
   const { Component, pageProps } = props;

  return (
    <Layout {...[some props from cookies]}>
      <Component {...pageProps} />
    </Layout>
  );
}



MyApp.getInitialProps = async (ctx) => {
  //need to access req.headers here
}

export default MyApp;
```

**Comment by normancapule** [comment]
Hello there! I came upon the same scenario as your question. I was able to make it play nicely with typescript with this:

```
MyApp.getInitialProps = async ({ Component, ctx }: AppContext) => {
... logic
}

```

**Reply by untitledlt** [reply]
That's not totally right.
```typescript
type AppContext = {
    Component: NextComponentType<NextPageContext, {}, {}>;
    AppTree: AppTreeType;
    ctx: NextPageContext;
    router: Router;
}
```

And
```typescript
interface NextPageContext {
    req?: IncomingMessage;
    res?: ServerResponse;
    // other props
}
```

But actually `req.cookies` contains cookies values so `req` is more like `NextApiRequest`
